hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java

```diff
@@ -748,6 +758,52 @@ public void recoverFile(OmKeyArgs keyArgs) throws IOException {
     ozoneClient.getProxy().getOzoneManagerClient().recoverKey(keyArgs, 0L);
   }
 
+  @Override
+  public long finalizeBlock(OmKeyLocationInfo block) throws IOException {
+    incrementCounter(Statistic.INVOCATION_FINALIZE_BLOCK, 1);
+    RpcClient rpcClient = (RpcClient) ozoneClient.getProxy();
+    XceiverClientFactory xceiverClientFactory = rpcClient.getXceiverClientManager();
+    Pipeline pipeline = block.getPipeline();
+    XceiverClientSpi client = null;
+    try {
+      // If pipeline is still open
+      if (pipeline.isOpen()) {
+        client = xceiverClientFactory.acquireClient(pipeline);
+        ContainerProtos.FinalizeBlockResponseProto finalizeBlockResponseProto =
+            ContainerProtocolCalls.finalizeBlock(client, block.getBlockID().getDatanodeBlockIDProtobuf(),
+                block.getToken());
+        return BlockData.getFromProtoBuf(finalizeBlockResponseProto.getBlockData()).getSize();
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to execute finalizeBlock command", e);
```

Comment:
Currently for all exception we are proceeding to get block length from DN. There may be case when container is still not CLOSED. 
I think we should get block length from DN when container replica is CLOSED.

Reply:
Checking CLOSED state replica is my initial idea too.  Then later I found there is more concise way.  We can leverage this getCommittedBlockLength call regardless of the replica state given that ratis is used to update the replica data. 

The implementation of getCommittedBlockLength compares the bcsid of involved block and replica's bcsid. If replica's bcsid is no less than block's bcsid, then the block info in this replica is a consensus result of raft, it's trustworthy.